pkg/uploadproxy/uploadproxy.go

```diff
@@ -234,22 +235,28 @@ func (app *uploadProxyApp) resolveUploadPath(pvcName, pvcNamespace, defaultPath
 		return "", err
 	}
 
-	contentType, found := pvc.Annotations[cc.AnnContentType]
-	if !found {
-		return defaultPath, nil
-	}
-
+	var path string
+	contentType, _ := pvc.Annotations[cc.AnnContentType]
 	switch contentType {
 	case string(cdiv1.DataVolumeKubeVirt), "":
-		return defaultPath, nil
+		path = defaultPath
 	case string(cdiv1.DataVolumeArchive):
 		if strings.Contains(defaultPath, "alpha") {
-			return common.UploadArchiveAlphaPath, nil
+			path = common.UploadArchiveAlphaPath
+		} else {
+			path = common.UploadArchivePath
 		}
-		return common.UploadArchivePath, nil
 	default:
 		return "", fmt.Errorf("rejecting upload request for PVC %s - upload content-type %s is invalid", pvcName, contentType)
 	}
+
+	//If using upload populator then the upload server will have PVC' name in the url
+	if populators.IsPVCDataSourceRefKind(pvc, cdiv1.UploadSourceRef) {
+		pvcPrimeName := populators.PVCPrimeName(pvc)
+		return app.urlResolver(pvcNamespace, pvcPrimeName, path), nil
+	}
+
+	return app.urlResolver(pvcNamespace, pvcName, path), nil
```

Comment:
as far as I see it the uploadproxy has to get the pvc prime in order to check the annotation and pod phase on the pvc. hence it has to know the pvc prime name anyways. Im not sure its so bad that it will know the format of getting pvc' name for that. you are also not revealing anything.
You are relying here on function createUploadServiceNameFromPvcName which is in the logic of the upload controller so I dont see how different it is between relying on the pvcprimename function.
If you update an annotation on the target pvc you rely on the annotation name that the populator updates so seems pretty much the same for me without the need to do an update on the target PVC and reconcile again to create the pvc'.

Reply:
I am pretty strongly against the upload proxy knowing anything about populator implementation details